tensorboard/plugins/histogram/histograms_plugin.py

```diff
@@ -128,45 +132,61 @@ def histograms_impl(self, tag, run, downsample_to=50):
-          Tensors.step AS step,
-          Tensors.data AS data,
-          Tensors.dtype AS dtype,
-          Tensors.shape AS shape
-        FROM Tensors
-        JOIN Tags
-          ON Tensors.series = Tags.tag_id
-        JOIN Runs
-          ON Tags.run_id = Runs.run_id 
-        WHERE
-          Runs.run_name = ?
-          AND Tags.tag_name = ?
-          AND Tags.plugin_name = ?
-          AND Tensors.step > -1
-      '''
-      if downsample_to is not None:
-        # Wrap the query in an outer one that samples.
-        query = '''
-          SELECT *
-          FROM 
-            (%(query)s
-             ORDER BY RANDOM()
-             LIMIT %(downsample_to)d)
-        ''' % {
-            'query': query,
-            'downsample_to': downsample_to,
-        }
-      query = '''
-        %s
-        ORDER BY step
-      ''' % query
-      cursor = db.execute(query, (run, tag, metadata.PLUGIN_NAME))
-      events = [(row[0], row[1], self._get_values(row[2], row[3], row[4]))
-                for row in cursor]
+      cursor = db.cursor()
+      # Prefetch the tag ID matching this run and tag.
+      cursor.execute(
+          '''
+          SELECT
+            tag_id
+          FROM Tags
+          JOIN Runs USING (run_id)
+          WHERE
+            Runs.run_name = :run
+            AND Tags.tag_name = :tag
+            AND Tags.plugin_name = :plugin
+          ''',
+          {'run': run, 'tag': tag, 'plugin': metadata.PLUGIN_NAME})
+      row = cursor.fetchone()
+      if not row:
+        raise ValueError('No histogram tag %r for run %r' % (tag, run))
+      (tag_id,) = row
+      # Fetch tensor values, optionally with linear-spaced sampling by step.
+      # For steps ranging from s_min to s_max and sample size k, this query
+      # divides the range into k - 1 equal-sized intervals and returns the
+      # lowest step at or above each of the k interval boundaries (which always
+      # includes s_min and s_max, and may be fewer than k results if there are
+      # intervals where no steps are present). For contiguous steps the results
+      # can be formally expressed as the following:
+      #   [s_min + math.ceil(i / k * (s_max - s_min)) for i in range(0, k + 1)]
+      cursor.execute(
+          '''
+          SELECT
+            MIN(step) AS step,
+            computed_time,
+            data,
+            dtype,
+            shape
+          FROM Tensors
+          INNER JOIN (
+            SELECT
+              MIN(step) AS min_step,
+              MAX(step) AS max_step
+            FROM Tensors
+            /* Filter out NULL so we can use TensorSeriesStepIndex. */
+            WHERE series = :tag_id AND step IS NOT NULL
+          )
+          /* Ensure we omit reserved rows, which have NULL step values. */
+          WHERE series = :tag_id AND step IS NOT NULL
+          /* Bucket rows into sample_size linearly spaced buckets, or do
+             no sampling if sample_size is NULL. */
+          GROUP BY
```

Comment:
If I'm understanding correctly, this `GROUP BY` makes it so that we only sample 1 event at each sampled step, right? ie, either the min or max per bucket that you had noted earlier.

Reply:
Yeah, the GROUP BY basically partitions the series by a "bucket index" so that we divide the range into k - 1 even sized intervals, and then within each interval we take the minimum aka first step in that interval.  The way we do the math, the max step is always assigned to its own singleton interval with bucket index k - 1, so it also get selected basically as the lower bound of a kth interval that just happens not to contain any other steps (so conceptually, I think of the selected steps as being the set of lower and upper bounds of k - 1 intervals instead).

Reply:
Note also that if the sample_size is null, we effectively set it to max_step - min_step + 1 which is the total number of distinct steps.  As it happens that means our formal now has (max_step - min_step) in both numerator and denominator, so they cancel and each step gets assigned a unique bucket index of step - min_step.

If sample_size is *greater* than the total number of steps, the bucket index for each step is assigned as m * step + C where m > 1, so each step will still get a unique bucket index and hence all steps will be selected.